### DISCUSSION TITLE
Mix git dependencies use pinned digest instead of the value from `ref`

### DISCUSSION THREAD

**Discussion by 1oglop1** [renovatebot/renovate#32815] [Request Help]
### How are you running Renovate?

A Mend.io-hosted app

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

latest master branch

### Please tell us more about your question or problem

This is a follow-up on https://github.com/renovatebot/renovate/discussions/31778

I think that by modifying the branch name or git tag name to `digest`, we lose a lot of information. Also, I do not know how common usage of `ref` is for branches and tags but renovate should support it regardless.

### Reproduction repo
-  https://github.com/jans-forks/renovate-reporoduction-31778 

This is the current behaviour

| Package | Update | Change |
|---|---|---|
| [git_lib_pkg](https://redirect.github.com/jans-forks/elixir_lib) | minor | `v0.2.0` -> `v0.3.0` |
| [lib_pkg](https://redirect.github.com/jans-forks/elixir_lib) | minor | `v0.1.0` -> `v0.3.0` |
| ref_lib_branch | digest | `main` -> `83b72d4` |
| ref_lib_pkg | digest | `4888c25` -> `83b72d4` |
| ref_lib_tag | digest | `v0.1.0` -> `83b72d4` |

Expected result

| Package | Update | Change |
|---|---|---|
| [git_lib_pkg](https://redirect.github.com/jans-forks/elixir_lib) | minor | `v0.2.0` -> `v0.3.0` |
| [lib_pkg](https://redirect.github.com/jans-forks/elixir_lib) | minor | `v0.1.0` -> `v0.3.0` |
| ref_lib_branch | digest | `main` -> `main` |
| ref_lib_pkg | digest | `4888c25` -> `83b72d4` |
| ref_lib_tag | digest | `v0.1.0` -> `v0.1.0` |

The deps `| ref_lib_branch | digest |` and `| ref_lib_tag | digest |` unexpectedly change in `mix.exs` to the digest from a readable value, while I think they should only change in `mix.lock`. Or we let the `mix` handle the lock file?

<img width="1134" alt="image" src="https://github.com/user-attachments/assets/9742dbaa-0efb-48db-86c4-d7bd2cec683a">
<img width="1123" alt="image" src="https://github.com/user-attachments/assets/1934f98d-c3da-430b-8ae8-a1b5108fbf96">

I tested with `pinDigests=false` but it did not help. However, when I updated the tests to include more cases, I noticed that
all `ref:` dependencies are recognized as a `git-tags` data source and the `updateTy. Is this a correct behaviour? There is also a `git-refs` datasource but forcing it via code, didn't help either.
Another thing is the `"updateType": "digest"` is always set for all `:ref` dependencies but not `:tag`.

Is this something I can try in tests or do I always need to run against the real repository?

<details>

```
"config": {
         "mix": [
           {
             "deps": [
               {
                 "depName": "lib_pkg",
                 "currentValue": "v0.1.0",
                 "datasource": "git-tags",
                 "packageName": "https://github.com/jans-forks/elixir_lib",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "v0.3.0",
                     "newValue": "v0.3.0",
                     "newMajor": 0,
                     "newMinor": 3,
                     "newPatch": 0,
                     "updateType": "minor",
                     "branchName": "renovate/mix"
                   }
                 ],
                 "versioning": "semver-coerced",
                 "warnings": [],
                 "sourceUrl": "https://github.com/jans-forks/elixir_lib",
                 "currentVersion": "v0.1.0",
                 "isSingleVersion": true,
                 "fixedVersion": "v0.1.0"
               },
               {
                 "depName": "git_lib_pkg",
                 "currentValue": "v0.2.0",
                 "datasource": "git-tags",
                 "packageName": "https://github.com/jans-forks/elixir_lib",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "v0.3.0",
                     "newValue": "v0.3.0",
                     "newMajor": 0,
                     "newMinor": 3,
                     "newPatch": 0,
                     "updateType": "minor",
                     "branchName": "renovate/mix"
                   }
                 ],
                 "versioning": "semver-coerced",
                 "warnings": [],
                 "sourceUrl": "https://github.com/jans-forks/elixir_lib",
                 "currentVersion": "v0.2.0",
                 "isSingleVersion": true,
                 "fixedVersion": "v0.2.0"
               },
               {
                 "depName": "ref_lib_pkg",
                 "currentDigest": "4888c25b27e47369296ebd66502a3d3ec6e955db",
                 "datasource": "git-tags",
                 "packageName": "https://github.com/jans-forks/elixir_lib",
                 "updates": [
                   {
                     "updateType": "digest",
                     "newDigest": "83b72d4940b13cc68e70a32aa14eb624b29b880a",
                     "branchName": "renovate/mix"
                   }
                 ],
                 "versioning": "semver-coerced",
                 "warnings": []
               },
               {
                 "depName": "ref_lib_tag",
                 "currentDigest": "v0.1.0",
                 "datasource": "git-tags",
                 "packageName": "https://github.com/jans-forks/elixir_lib",
                 "updates": [
                   {
                     "updateType": "digest",
                     "newDigest": "83b72d4940b13cc68e70a32aa14eb624b29b880a",
                     "branchName": "renovate/mix"
                   }
                 ],
                 "versioning": "semver-coerced",
                 "warnings": []
               },
               {
                 "depName": "ref_lib_branch",
                 "currentDigest": "main",
                 "datasource": "git-tags",
                 "packageName": "https://github.com/jans-forks/elixir_lib",
                 "updates": [
                   {
                     "updateType": "digest",
                     "newDigest": "83b72d4940b13cc68e70a32aa14eb624b29b880a",
                     "branchName": "renovate/mix"
                   }
                 ],
                 "versioning": "semver-coerced",
                 "warnings": []
               }
             ],
             "lockFiles": ["el_app/mix.lock"],
             "packageFile": "el_app/mix.exs"
           }
         ]
       }
  ```

</details>

  
I wonder if this behaviour is somehow caused by this code, which sets both `currentValue` and `currentDigest`.
  https://github.com/renovatebot/renovate/blob/d76b14a93a34edff5abe087410ee38b67b063d85/lib/modules/manager/mix/extract.ts#L50-L56

**Comment by paulo-ferraz-oliveira** [comment]
I agree that for

```elixir
      {<atom>, <opts> git: "...", ref: "<TAG>"},
      {<atom>, <opts> git: "...", ref: "<DIGEST>"},
      {<atom>, <opts> git: "...", ref: "<BRANCH>"}
```

`<BRANCH>` should not be updated to a digest in the `mix.exs` file, especially since Elixir will/should update the `mix.lock` equivalent, and, as you state, you lose information for the (human) consumer of the file.

I don't see how `<DIGEST>` can be **safely** updated, but I don't use it that much. (when I do it's because I want to pin to a specific element and **never** move that reference)

As you kinda state, @1oglop1, on the other hand, if you're using `<TAG>` you should probably be using the `tag:` option; the same for `<BRANCH>`, by using the `branch:` option, so this almost approaches an edge case, apart from the digest update.

**Reply by 1oglop1** [reply]
good point, I think the `<DIGEST>` should not be updated by renovate. 
If you are pinning your dependency to a certain commit, the intent is not to update IMO.
And therefore the manager should mark such dependency as skip. There is also a skip reason type https://github.com/renovatebot/renovate/blob/cec0a0b1d1da6f6e4a079ec8a4272c4511f563d2/lib/types/skip-reason.ts#L1
So we could have this behaviour:
`ref: <DIGEST>` - skip - and let `mix lock` handle it
`ref: <TAG>` - use strategy as `:tag`
`ref: <BRANCH>` - use the same strategy as `:branch` 
And avoid pining the digest entirely. By pinning I mean the transformation of `tag` -> `digest`.
That's why rarkins asked to try `pinDigests: false` - but it did not have the expected effect, so I'll need to dig deeper to find out where it happens.

**Reply by rarkins** [reply]
If a digest is in the package file (not lock file) then Renovate should update it. Users can disable such updates if they wish, but it should not be disabled by default. It's a typical use case that:
- People need to depend on a git repo
- They want some reproducibility, so pin it to a hash on main branch
- They want Renovate to update the hash

**Reply by 1oglop1** [reply]
Thanks for clarifying @rarkins it probably makes sense not to disable `ref` updates by default.
Do you agree that the branch and the tag should not be changed to `digest` in the package file when represented as `ref`? 
For the implementation of this, I am not sure if there is any `datasource` that can make sure it works this way or if the `manager` logic is enough. 
I.E. manager sets the `git-*` datasource eg.

```ts
// ref is not digest
const depNotDigest: PackageDependency = {
            currentValue: branchOrTag,
            datasource: git ? GitTagsDatasource.id : GithubTagsDatasource.id,
            ...
          };
// ref is digest
const depDigest: PackageDependency = {
            currentValue: branchOrTag,
            currentDigest: ref,
            datasource: git ? GitTagsDatasource.id : GithubTagsDatasource.id,
            ...
          };
```
If my theory is correct, setting `currentDigest` will cause renovate to replace `ref` inside the package file with `digest`.
Are there any tests that I can read to support my theory or is it only possible to test on the real repo?

**Reply by rarkins** [reply]
> Do you agree that the branch and the tag should not be changed to `digest` in the package file when represented as `ref`?

If there's a branch or tag, they should not be changed to a digest unless pinDigests=true. If pinDigests=false then branches/tags should only be updated to another branch/tag
